source/kaleidic/lubeck2.d

```diff
@@ -71,87 +71,411 @@ unittest
+        Slice!(const(A)*, 2, kindA) a
+    )
+        if (is(Unqual!A == Unqual!B))
+    in
+    {
+        assert(a.length!0 == b.length);
+    }
+    do
+    {
+        auto scopeB = b.lightScope.lightConst;
+        return .mtimes!assumedMatrix(scopeB, a);
+    }
+}
+
 /++
-General matrix-matrix multiplication. Allocates result to using Mir refcounted arrays.
 Params:
-a = m(rows) x k(cols) matrix
-b = k(rows) x n(cols) matrix
-Result: 
-m(rows) x n(cols)
+    a = 1(rows) x n(cols) vector
+    b = n(rows) x 1(cols) vector
+Result:
+    dot product
 +/
 @safe pure nothrow @nogc
-Slice!(RCI!T, 2) mtimes(T, SliceKind kindA, SliceKind kindB)(
-    Slice!(const(T)*, 2, kindA) a,
-    Slice!(const(T)*, 2, kindB) b
+Unqual!A mtimes(A, B, SliceKind kindA, SliceKind kindB)(
+    Slice!(const(A)*, 1, kindA) a,
+    Slice!(const(B)*, 1, kindB) b
 )
-    if (isFloatingPoint!T || isComplex!T)
+    if (is(Unqual!A == Unqual!B))
 in
 {
-    assert(a.length!1 == b.length!0);
-}
-out (c)
-{
-    assert(c.length!0 == a.length!0);
-    assert(c.length!1 == b.length!1);
+    assert(a.length == b.length);
 }
 do
 {
-    // optimisations for spcecial cases can be added in the future
-    auto c = mininitRcslice!T(a.length!0, b.length!1);
-    gemm(cast(T)1, a, b, cast(T)0, c.lightScope);
-    return c;
+    import mir.blas: dot;
+
+    return dot(a, b);
 }
 
 /// ditto
 @safe pure nothrow @nogc
-Slice!(RCI!(Unqual!A), 2) mtimes(A, B, SliceKind kindA, SliceKind kindB)(
-    auto ref const Slice!(RCI!A, 2, kindA) a,
-    auto ref const Slice!(RCI!B, 2, kindB) b
+Unqual!A mtimes(A, B, SliceKind kindA, SliceKind kindB)(
+    auto ref const Slice!(RCI!A, 1, kindA) a,
+    auto ref const Slice!(RCI!B, 1, kindB) b
 )
     if (is(Unqual!A == Unqual!B))
 in
 {
-    assert(a.length!1 == b.length!0);
+    assert(a.length == b.length);
 }
 do
 {
     auto scopeA = a.lightScope.lightConst;
     auto scopeB = b.lightScope.lightConst;
-    return .mtimes(scopeA, scopeB);
+    return mtimes(scopeA, scopeB);
 }
 
+/// ditto
 @safe pure nothrow @nogc
-Slice!(RCI!(Unqual!A), 2) mtimes(A, B, SliceKind kindA, SliceKind kindB)(
-    auto ref const Slice!(RCI!A, 2, kindA) a,
-    Slice!(const(B)*, 2, kindB) b
+Unqual!A mtimes(A, B, SliceKind kindA, SliceKind kindB)(
+    auto ref const Slice!(RCI!A, 1, kindA) a,
+    Slice!(const(B)*, 1, kindB) b
 )
-if (is(Unqual!A == Unqual!B))
+    if (is(Unqual!A == Unqual!B))
 in
 {
-    assert(a.length!1 == b.length!0);
+    assert(a.length == b.length);
 }
 do
 {
     auto scopeA = a.lightScope.lightConst;
-    return .mtimes(scopeA, b);
+    return mtimes(scopeA, b);
 }
 
+/// ditto
 @safe pure nothrow @nogc
-Slice!(RCI!(Unqual!A), 2) mtimes(A, B, SliceKind kindA, SliceKind kindB)(
-    Slice!(const(A)*, 2, kindA) a,
-    auto ref const Slice!(RCI!B, 2, kindB) b
+Unqual!A mtimes(A, B, SliceKind kindA, SliceKind kindB)(
+    Slice!(const(A)*, 1, kindA) a,
+    auto ref const Slice!(RCI!B, 1, kindB) b
```

Comment:
GitHub doesn't show a proper diff. Please rearrange the new API at the end of the mtimes, block for a simpler review.